rclrs/src/node/graph.rs

```diff
@@ -0,0 +1,568 @@
+        ) -> rcl_ret_t,
+    ) -> Result<TopicNamesAndTypes, RclrsError> {
+        let mut rcl_names_and_types = unsafe { rmw_get_zero_initialized_names_and_types() };
+        let node_name = CString::new(node).unwrap();
+        let node_namespace = CString::new(namespace).unwrap();
+        unsafe {
+            getter(
+                &*self.rcl_node_mtx.lock() as *const _,
+                &mut rcutils_get_default_allocator(),
+                node_name.as_ptr(),
+                node_namespace.as_ptr(),
+                &mut rcl_names_and_types,
+            )
+        };
+
+        convert_names_and_types(rcl_names_and_types)
+    }
+
+    /// Get publisher or subscriber info by topic
+    fn get_publisher_subscriber_info_by_topic(
+        &self,
+        topic: &str,
+        getter: unsafe extern "C" fn(
+            *const rcl_node_t,
+            *mut rcl_allocator_t,
+            *const ::std::os::raw::c_char,
+            bool,
+            *mut rcl_topic_endpoint_info_array_t,
+        ) -> rcl_ret_t,
+    ) -> Result<Vec<TopicEndpointInfo>, RclrsError> {
+        let topic = CString::new(topic).unwrap();
+        let mut rcl_publishers_info =
+            unsafe { rmw_get_zero_initialized_topic_endpoint_info_array() };
+
+        unsafe {
+            getter(
+                &*self.rcl_node_mtx.lock(),
+                &mut rcutils_get_default_allocator() as *mut _,
+                topic.as_ptr(),
+                true,
+                &mut rcl_publishers_info as *mut _,
+            )
+            .ok()?
+        }
+
+        let topic_endpoint_infos_slice = unsafe {
+            slice::from_raw_parts(rcl_publishers_info.info_array, rcl_publishers_info.size)
+        };
+
+        let topic_endpoint_infos_vec = topic_endpoint_infos_slice
+            .iter()
+            .map(|info| {
+                let (node_name, node_namespace, topic_type, endpoint_type) = unsafe {
+                    (
+                        CStr::from_ptr(info.node_name)
+                            .to_string_lossy()
+                            .into_owned(),
+                        CStr::from_ptr(info.node_namespace)
+                            .to_string_lossy()
+                            .into_owned(),
+                        CStr::from_ptr(info.topic_type)
+                            .to_string_lossy()
+                            .into_owned(),
+                        info.endpoint_type,
+                    )
+                };
+                TopicEndpointInfo {
+                    node_name,
+                    node_namespace,
+                    topic_type,
+                    endpoint_type,
+                }
+            })
+            .collect();
+
+        Ok(topic_endpoint_infos_vec)
+    }
+}
+
+/// Convert a rmw_names_and_types_t object to a HashMap
+fn convert_names_and_types(
+    rcl_names_and_types: rmw_names_and_types_t,
+) -> Result<HashMap<String, Vec<String>>, RclrsError> {
```

Comment:
Nit: this function doesn't need to return a `Result`, so I'd prefer to add `Ok()` to each caller.